### PR TITLE
Auto create valid extensions

### PR DESCRIPTION
**Pull request by graingert**

### CODE REVIEW

File: src/format_cef/cef.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    ensure_in_range,
+    escaper,
+    float_sanitiser,
+    format_cef,
+    int_sanitiser,
+    str_sanitiser,
+)
+from ._cef.base import valid_extensions as _valid_extensions
+
+valid_extensions = dict(_valid_extensions)
```
Comment (funkyHat) [inline src/format_cef/cef.py]: ? it's already a dict, what is this for?

Comment (graingert) [reply src/format_cef/cef.py]: to stop consumers mutating valid_extensions, a poor imitation of MappingProxyType

eventually I want to move everything out of this and have it only export format_cef publicly

File: gen_extensions.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                    if re.match(
+                        r"\A.*numbers are( between)? 0 (?(1)and|to) 65535.\Z",
+                        record[MEANING],
+                    )
+                    else map_optional(int, record[LENGTH] or None)
+                ),
+                DATA_TYPE: record[DATA_TYPE],
+            }
+            for record in functools.reduce(
+                concat_ophans_and_widows,
```
Comment (graingert) [inline gen_extensions.py]:
```suggestion
                concat_orphans_and_widows,
```

File: gen_extensions.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+drop_all_whitespace_cols = [CEF_KEY_NAME, FULL_NAME, LENGTH]
+cef_fieldnames = (CEF_KEY_NAME, FULL_NAME, DATA_TYPE, LENGTH, MEANING)
+
+
+def map_optional(fn, v):
+    return None if v is None else fn(v)
+
+
+def process(camelot_tables):
+    def concat_ophans_and_widows(acc, v):
```
Comment (graingert) [inline gen_extensions.py]:
```suggestion
    def concat_orphans_and_widows(acc, v):
```

File: tests/test_cef.py [hunk 4]
```diff
@@ -79,17 +94,43 @@ def test_format_cef(self):
...
-            r"10|act=explode \= !",
+            six.ensure_binary(
+                r"CEF:0|acme corp|TNT|1.0|404 \| not found|Explosives not found|"
+                r"10|act=explode \= !"
+            ),
         )
 
     def test_extensions_with_prototypical_data(self):
         # This pretty much just checks that all the sanitisers execute!
+        def legacy_ip():
```
Comment (funkyHat) [inline tests/test_cef.py]: unused?

Comment (graingert) [reply tests/test_cef.py]: dangit

**Pull request merged**